### ISSUE TITLE
Add way to de-duplicate function calls

### ISSUE THREAD

**Issue by chrisbanes** [Kotlin/kotlinx.coroutines#1097]
I currently have a number of places where I request a specific URL. It would be nice if any concurrent requests would be de-duplicated so that any requested which are started while one is in progress, joins the other and receives the same result.

I imagine that this would be similar to Go's [singleflight](https://godoc.org/golang.org/x/sync/singleflight): a collection wrapper around a Mutex, keyed on a string.

**Comment by elizarov** [comment]
Does it really has to be keyed by a string? We are not in Go, you know, we can do much better.

**Comment by chrisbanes** [comment]
Yeah agreed, although I do like the simplicity of a string. Using an `inline class Key(val k: String)` would work just as well though.

**Comment by LouisCAD** [comment]
I have to link this gist from @objcode, because it might help folks having this call de-duplicating use case: https://gist.github.com/objcode/7ab4e7b1df8acd88696cb0ccecad16f7#file-concurrencyhelpers-kt-L162-L339
